// @pinecone-database/pinecone/dist/pinecone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pinecone = void 0;
var pinecone_generated_ts_fetch_1 = require("./pinecone-generated-ts-fetch");
var control_1 = require("./control");
var errors_1 = require("./errors");
var middleware_1 = require("./utils/middleware");
var data_1 = require("./data");
var validator_1 = require("./validator");
var utils_1 = require("./utils");
/**
 * The `Pinecone` class is the main entrypoint to this sdk. You will use
 * instances of it to create and manage indexes as well as perform data
 * operations on those indexes after they are created.
 *
 * ### Initializing the client
 *
 * There are two pieces of configuration required to use the Pinecone client: an API key and environment value. These values can be passed using environment variables or in code through a configuration object. Find your configuration values in the console dashboard at [https://app.pinecone.io](https://app.pinecone.io)
 *
 * ### Using environment variables
 *
 * The environment variables used to configure the client are the following:
 *
 * ```bash
 * export PINECONE_API_KEY="your_api_key"
 * export PINECONE_ENVIRONMENT="your_environment"
 * ```
 *
 * When these environment variables are set, the client constructor does not require any additional arguments.
 *
 * ```typescript
 * import { Pinecone } from '@pinecone-database/pinecone';
 *
 * const pinecone = new Pinecone();
 * ```
 *
 * ### Using a configuration object
 *
 * If you prefer to pass configuration in code, the constructor accepts a config object containing the `apiKey` and `environment` values. This
 * could be useful if your application needs to interact with multiple projects, each with a different configuration.
 *
 * ```typescript
 * import { Pinecone } from '@pinecone-database/pinecone';
 *
 * const pinecone = new Pinecone({
 *   apiKey: 'your_api_key',
 *   environment: 'your_environment',
 * });
 *
 * ```
 *
 * See {@link PineconeConfiguration} for a full description of available configuration options.
 */
var Pinecone = /** @class */ (function () {
    /**
     * @example
     * ```
     * import { Pinecone } from '@pinecone-database/pinecone';
     *
     * const pinecone = new Pinecone({
     *  apiKey: 'my-api-key',
     *  environment: 'us-west1-gcp'
     * });
     * ```
     *
     * @constructor
     * @param options - The configuration options for the pinecone.
     */
    function Pinecone(options) {
        if (options === undefined) {
            options = this._readEnvironmentConfig();
        }
        this._validateConfig(options);
        this.config = options;
        var apiKey = options.apiKey, environment = options.environment;
        var controllerPath = "https://controller.".concat(environment, ".pinecone.io");
        var apiConfig = {
            basePath: controllerPath,
            apiKey: apiKey,
            queryParamsStringify: utils_1.queryParamsStringify,
            headers: {
                'User-Agent': (0, utils_1.buildUserAgent)(false),
            },
            fetchApi: (0, utils_1.getFetch)(options),
            middleware: middleware_1.middleware,
        };
        var api = new pinecone_generated_ts_fetch_1.IndexOperationsApi(new pinecone_generated_ts_fetch_1.Configuration(apiConfig));
        this._configureIndex = (0, control_1.configureIndex)(api);
        this._createCollection = (0, control_1.createCollection)(api);
        this._createIndex = (0, control_1.createIndex)(api);
        this._describeCollection = (0, control_1.describeCollection)(api);
        this._deleteCollection = (0, control_1.deleteCollection)(api);
        this._describeIndex = (0, control_1.describeIndex)(api);
        this._deleteIndex = (0, control_1.deleteIndex)(api);
        this._listCollections = (0, control_1.listCollections)(api);
        this._listIndexes = (0, control_1.listIndexes)(api);
    }
    /**
     * @internal
     * This method is used by {@link Pinecone.constructor} to read configuration from environment variables.
     *
     * It looks for the following environment variables:
     * - `PINECONE_ENVIRONMENT`
     * - `PINECONE_API_KEY`
     * - `PINECONE_PROJECT_ID`
     *
     * @returns A {@link PineconeConfiguration} object populated with values found in environment variables.
     */
    Pinecone.prototype._readEnvironmentConfig = function () {
        if (typeof process === 'undefined' || !process || !process.env) {
            throw new errors_1.PineconeEnvironmentVarsNotSupportedError('Your execution environment does not support reading environment variables from process.env, so a configuration object is required when calling new Pinecone()');
        }
        var environmentConfig = {};
        var requiredEnvVarMap = {
            environment: 'PINECONE_ENVIRONMENT',
            apiKey: 'PINECONE_API_KEY',
        };
        var missingVars = [];
        for (var _i = 0, _a = Object.entries(requiredEnvVarMap); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], envVar = _b[1];
            var value = process.env[envVar] || '';
            if (!value) {
                missingVars.push(envVar);
            }
            environmentConfig[key] = value;
        }
        if (missingVars.length > 0) {
            throw new errors_1.PineconeConfigurationError("Since you called 'new Pinecone()' with no configuration object, we attempted to find client configuration in environment variables but the required environment variables were not set. Missing variables: ".concat(missingVars.join(', '), "."));
        }
        var optionalEnvVarMap = { projectId: 'PINECONE_PROJECT_ID' };
        for (var _c = 0, _d = Object.entries(optionalEnvVarMap); _c < _d.length; _c++) {
            var _e = _d[_c], key = _e[0], envVar = _e[1];
            var value = process.env[envVar];
            if (value !== undefined) {
                environmentConfig[key] = value;
            }
        }
        return environmentConfig;
    };
    /**
     * Describe a Pinecone index
     *
     * @example
     * ```js
     * const indexConfig = await pinecone.describeIndex('my-index')
     * console.log(indexConfig)
     * // {
     * //    database: {
     * //      name: 'my-index',
     * //      metric: 'cosine',
     * //      dimension: 256,
     * //      pods: 2,
     * //      replicas: 2,
     * //      shards: 1,
     * //      podType: 'p1.x2',
     * //      metadataConfig: { indexed: [Array] }
     * //    },
     * //    status: { ready: true, state: 'Ready' }
     * // }
     * ```
     *
     * @param indexName - The name of the index to describe.
     * @returns A promise that resolves to {@link IndexMeta}
     */
    Pinecone.prototype.describeIndex = function (indexName) {
        return this._describeIndex(indexName);
    };
    /**
     * List all Pinecone indexes
     * @example
     * ```js
     * const indexes = await pinecone.listIndexes()
     * console.log(indexes)
     * // [ 'my-index', 'my-other-index' ]
     * ```
     *
     * @returns A promise that resolves to an array of index names
     */
    Pinecone.prototype.listIndexes = function () {
        return this._listIndexes();
    };
    /**
     * Creates a new index.
     *
     * @example
     * The minimum required configuration to create an index is the index name and dimension.
     * ```js
     * await pinecone.createIndex({ name: 'my-index', dimension: 128 })
     * ```
     * @example
     * In a more expansive example, you can specify the metric, number of pods, number of replicas, and pod type.
     * ```js
     * await pinecone.createIndex({
     *  name: 'my-index',
     *  dimension: 1536,
     *  metric: 'cosine',
     *  pods: 1,
     *  replicas: 2,
     *  podType: 'p1.x1'
     * })
     * ```
     *
     * @example
     * If you would like to create the index only if it does not already exist, you can use the `suppressConflicts` boolean option.
     * ```js
     * await pinecone.createIndex({
     *   name: 'my-index',
     *   dimension: 1536,
     *   suppressConflicts: true
     * })
     * ```
     *
     * @example
     * If you plan to begin upserting immediately after index creation is complete, you should use the `waitUntilReady` option. Otherwise, the index may not be ready to receive data operations when you attempt to upsert.
     * ```js
     * await pinecone.createIndex({
     *  name: 'my-index',
     *  dimension: 1536,
     *  waitUntilReady: true
     * });
     *
     * const records = [
     *   // PineconeRecord objects with your embedding values
     * ]
     * await pinecone.index('my-index').upsert(records)
     * ```
     *
     * @example
     * By default all metadata fields are indexed when records are upserted with metadata, but if you want to improve performance you can specify the specific fields you want to index. This example is showing a few hypothetical metadata fields, but the values you'd use depend on what metadata you plan to store with records in your Pinecone index.
     * ```js
     * await pinecone.createIndex({
     *   name: 'my-index',
     *   dimension: 1536,
     *   metadataConfig: { 'indexed' : ['productName', 'productDescription'] }
     * })
     * ```
     *
     * @param options - The index configuration.
     *
     * @see [Distance metrics](https://docs.pinecone.io/docs/indexes#distance-metrics)
     * @see [Pod types and sizes](https://docs.pinecone.io/docs/indexes#pods-pod-types-and-pod-sizes)
     * @throws {@link Errors.PineconeArgumentError} when invalid arguments are provided.
     * @throws {@link Errors.PineconeConflictError} when attempting to create an index using a name that already exists in your project.
     * @throws {@link Errors.PineconeBadRequestError} when index creation fails due to invalid parameters being specified or other problem such as project quotas limiting the creation of any additional indexes.
     *
     * @returns A promise that resolves when the request to create the index is completed. Note that the index is not immediately ready to use. You can use the `describeIndex` function to check the status of the index.
     */
    Pinecone.prototype.createIndex = function (options) {
        return this._createIndex(options);
    };
    /**
     * Deletes an index
     *
     * @example
     * ```js
     * await pinecone.deleteIndex('my-index')
     * ```
     *
     * @param indexName - The name of the index to delete.
     * @returns A promise that resolves when the request to delete the index is completed.
     * @throws {@link Errors.PineconeArgumentError} when invalid arguments are provided
     */
    Pinecone.prototype.deleteIndex = function (indexName) {
        return this._deleteIndex(indexName);
    };
    /**
     * Configure an index
     *
     * Use this method to update configuration on an existing index. You can update the number of pods, replicas, and pod type. You can also update the metadata configuration.
     *
     * @example
     * ```js
     * await pinecone.configureIndex('my-index', { replicas: 2, podType: 'p1.x2' })
     * ```
     *
     * @param indexName - The name of the index to configure.
     * @param options - The configuration properties you would like to update
     */
    Pinecone.prototype.configureIndex = function (indexName, options) {
        return this._configureIndex(indexName, options);
    };
    /**
     * Create a new collection from an existing index
     *
     * @example
     * ```js
     * const indexList = await pinecone.listIndexes()
     * await pinecone.createCollection({
     *  name: 'my-collection',
     *  source: indexList[0]
     * })
     * ```
     *
     *
     * @param options - The collection configuration.
     * @param options.name - The name of the collection. Must be unique within the project and contain alphanumeric and hyphen characters. The name must start and end with alphanumeric characters.
     * @param options.source - The name of the index to use as the source for the collection.
     * @returns a promise that resolves when the request to create the collection is completed.
     */
    Pinecone.prototype.createCollection = function (options) {
        return this._createCollection(options);
    };
    /**
     * List all collections in a project
     *
     * @example
     * ```js
     * await pinecone.listCollections()
     * ```
     *
     * @returns A promise that resolves to an array of collection objects.
     */
    Pinecone.prototype.listCollections = function () {
        return this._listCollections();
    };
    /**
     * Delete a collection by collection name
     *
     * @example
     * ```
     * const collectionList = await pinecone.listCollections()
     * const collectionName = collectionList[0]
     * await pinecone.deleteCollection(collectionName)
     * ```
     *
     * @param collectionName - The name of the collection to delete.
     * @returns A promise that resolves when the request to delete the collection is completed.
     */
    Pinecone.prototype.deleteCollection = function (collectionName) {
        return this._deleteCollection(collectionName);
    };
    /**
     * Describe a collection
     *
     * @example
     * ```js
     * await pinecone.describeCollection('my-collection')
     * ```
     *
     * @param collectionName - The name of the collection to describe.
     * @returns A promise that resolves to a collection object with type {@link CollectionDescription}.
     */
    Pinecone.prototype.describeCollection = function (collectionName) {
        return this._describeCollection(collectionName);
    };
    /** @internal */
    Pinecone.prototype._validateConfig = function (options) {
        (0, validator_1.buildValidator)('The client configuration', data_1.PineconeConfigurationSchema)(options);
    };
    /**
     * @returns The configuration object that was passed to the Pinecone constructor.
     */
    Pinecone.prototype.getConfig = function () {
        return this.config;
    };
    /**
     * Targets a specific index for performing data operations.
     *
     * ```typescript
     * import { Pinecone } from '@pinecone-database/pinecone';
     *
     * const pinecone = new Pinecone()
     * const index = pinecone.index('index-name')
     * ```
     *
     * #### Targeting an index, with user-defined Metadata types
     *
     * If you are storing metadata alongside your vector values inside your Pinecone records, you can pass a type parameter to `index()` in order to get proper TypeScript typechecking when upserting and querying data.
     *
     * ```typescript
     * const pinecone = new Pinecone();
     *
     * type MovieMetadata = {
     *   title: string,
     *   runtime: numbers,
     *   genre: 'comedy' | 'horror' | 'drama' | 'action'
     * }
     *
     * // Specify a custom metadata type while targeting the index
     * const index = pinecone.index<MovieMetadata>('test-index');
     *
     * // Now you get type errors if upserting malformed metadata
     * await index.upsert([{
     *   id: '1234',
     *   values: [
     *     .... // embedding values
     *   ],
     *   metadata: {
     *     genre: 'Gone with the Wind',
     *     runtime: 238,
     *     genre: 'drama',
     *
     *     // @ts-expect-error because category property not in MovieMetadata
     *     category: 'classic'
     *   }
     * }])
     *
     * const results = await index.query({
     *    vector: [
     *     ... // query embedding
     *    ],
     *    filter: { genre: { '$eq': 'drama' }}
     * })
     * const movie = results.matches[0];
     *
     * if (movie.metadata) {
     *   // Since we passed the MovieMetadata type parameter above,
     *   // we can interact with metadata fields without having to
     *   // do any typecasting.
     *   const { title, runtime, genre } = movie.metadata;
     *   console.log(`The best match in drama was ${title}`)
     * }
     * ```
     *
     * @typeParam T - The type of metadata associated with each record.
     * @param indexName - The name of the index to target.
     * @typeParam T - The type of the metadata object associated with each record.
     * @returns An {@link Index} object that can be used to perform data operations.
     */
    Pinecone.prototype.index = function (indexName) {
        return new data_1.Index(indexName, this.config);
    };
    /**
     * {@inheritDoc index}
     */
    // Alias method to match the Python SDK capitalization
    Pinecone.prototype.Index = function (indexName) {
        return this.index(indexName);
    };
    return Pinecone;
}());
exports.Pinecone = Pinecone;
